Channel participation: implement remove channel for application channels

wlahti opened the pull request:
#### Type of change

- New feature
- Test update

#### Description

This stacks on top of https://github.com/hyperledger/fabric/pull/1628. It implements the registrar's RemoveChannel function and includes a basic integration test.

#### Related issues

[FAB-17964](https://jira.hyperledger.org/browse/FAB-17964) - implement registrar RemoveChannel
[FAB-18093](https://jira.hyperledger.org/browse/FAB-18093) - integration test

joshhus commented:
README changes not intended as part of this; have to check my sync.

wlahti committed:
Channel participation: implement RemoveChannel for app chan

FAB-17964

Signed-off-by: Will Lahti <wtlahti@us.ibm.com>

wlahti committed:
Channel participation: basic RemoveChannel integration test

FAB-18093

Signed-off-by: Will Lahti <wtlahti@us.ibm.com>

tock-ibm commented on orderer/common/multichannel/registrar.go:
```diff
@@ -581,11 +581,42 @@ func (r *Registrar) joinAsFollower(ledgerRes *ledgerResources, clusterConsenter
+	cs, ok := r.chains[channelID]
+	if !ok {
+		return types.ErrChannelNotExist
+	}
```
a follower may also be removed, not just a chain.

wlahti replied:
Done. Haven't added the tests yet (just noticed your join channel as follower PR was merged so I'll use that as a reference).

wlahti replied:
Added the UT for this (a while ago).

tock-ibm commented on orderer/common/multichannel/registrar.go:
```diff
@@ -581,11 +581,42 @@ func (r *Registrar) joinAsFollower(ledgerRes *ledgerResources, clusterConsenter
+
+	raftChain.HaltExternal()
+
+	err := r.ledgerFactory.Remove(channelID)
+	if err != nil {
+		return errors.Errorf("error removing ledger for channel %s", channelID)
+	}
```
This may take a long time for a big chain, and blocks the request thread. We need to benchmark the Remove  and see if it is a real issue. If it is, put a Jira out to implement this asynch to the calling thread. That is, spin a go-routine to do it, and in the meantime put the channel in a "quarantine" map. In quarantine, a channel cannot be joined or deleted again until it finishes deleting its storage.

tock-ibm commented on orderer/common/multichannel/registrar.go:
```diff
@@ -581,11 +581,42 @@ func (r *Registrar) joinAsFollower(ledgerRes *ledgerResources, clusterConsenter
+		return errors.Errorf("cannot remove non-raft application channel %s", channelID)
+	}
+
+	raftChain.HaltExternal()
```
Not sure we need `HaltExternal`, I think a simple fix in the raft chain can let us use `Halt`.

lindluni commented:
/azp run

tock-ibm commented on orderer/common/multichannel/registrar_test.go:
```diff
@@ -806,6 +805,96 @@ func TestRegistrar_RemoveChannel(t *testing.T) {
+			err = registrar.RemoveChannel("my-raft-channel", false)
+			require.NoError(t, err)
+			// After deleting the chain, it no longer exists
+			require.Nil(t, registrar.GetChain("my-raft-channel"))
```
Check that ledger resources were removed as well.

wlahti replied:
Done

tock-ibm commented on orderer/common/multichannel/util_test.go:
```diff
@@ -31,6 +31,10 @@ func (c *mockChainCluster) StatusReport() (types.ClusterRelation, types.Status)
+func (c *mockChainCluster) HaltExternal() {
+	c.mockChain.Halt()
+}
+
```
Why is this needed?

wlahti replied:
Good catch. This is leftover from the previous implementation in the halt PR.  I'll remove this.

tock-ibm commented on orderer/common/multichannel/registrar.go:
```diff
@@ -577,8 +577,71 @@ func (r *Registrar) joinAsFollower(ledgerRes *ledgerResources, clusterConsenter
 		return types.ErrSystemChannelExists
 	}
-	return errors.New("Not implemented yet")
+
+	cs, ok := r.chains[channelID]
```
Add a check right above here that rejects the removal of the system channel, and returns "not yet implemented". Removal of the system channel is a different (Jira) story that requires a special process, since it mandates the conversion of every chain life-cycle from etcdraft+inactive to etcdraft+follower. This may require a restart of the orderer or the restart of every chain, as we'll decide when we tackle that story.

wlahti replied:
Done

tock-ibm commented on orderer/common/multichannel/registrar.go:
```diff
@@ -577,8 +577,71 @@ func (r *Registrar) joinAsFollower(ledgerRes *ledgerResources, clusterConsenter
+		return errors.Errorf("cannot remove non-raft application channel %s", channelID)
+	}
+
+	return nil
```
I am thinking, there is nothing in the "Remove" that is specific to raft/follower... We call `Halt`, we remove the ledger, and that is it (unlike the "Join"...). So perhaps we can make the code simpler by removing this check?

wlahti replied:
Done

tock-ibm commented on orderer/common/multichannel/registrar.go:
```diff
@@ -577,8 +577,71 @@ func (r *Registrar) joinAsFollower(ledgerRes *ledgerResources, clusterConsenter
+	err = r.ledgerFactory.Remove(channelID)
+	if err != nil {
+		return errors.Errorf("error removing ledger for channel %s", channelID)
+	}
```
I am worried because this operation is performed under the global registrar lock. If this takes even more then a fraction of a second, it will impact the performance of the server on all other channels. This method 
```go

// GetChain retrieves the chain support for a chain if it exists.
func (r *Registrar) GetChain(chainID string) *ChainSupport {
	r.lock.RLock()
	defer r.lock.RUnlock()

	return r.chains[chainID]
}
```
is called by every client request, through `BroadcastChannelSupport` so we can't really hold that lock for too long (for write - exclusive).

What I propose is async deletion: 
- introduce a `toBeRemoved` set (e.g. `map[string]bool`) that holds channels that are being removed. 
- After stopping the channel move the channelID to that set, 
- and only then delete the ledger, but make the delete async with a different go-routine.
- release the lock.
- when the delete go-routine finishes, it removes the channelID from the `toBeRemoved`
- any attempt to Join or Remove should also check the `toBeRemoved`

What do you think? @wlahti @jyellick @yacovm 
We can possibly do that in another Jira that will improve upon this one.

tock-ibm replied:
Opened https://jira.hyperledger.org/browse/FAB-18132 for this.

tock-ibm commented on orderer/common/multichannel/registrar.go:
```diff
@@ -577,8 +577,71 @@ func (r *Registrar) joinAsFollower(ledgerRes *ledgerResources, clusterConsenter
 func (r *Registrar) RemoveChannel(channelID string, removeStorage bool) error {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+
```
This must be a Lock / Unlock as we are writing the maps here, and need exclusive access.
This means that you can't use method `r.SystemChannelID()` since locks in go are not re-entrant.

wlahti replied:
Done

wlahti committed:
Replace Height with BlockNum in raft IT helper assertBlockReception

This helper function is checking for the block number and not
the height. Rename to avoid confusion.

Signed-off-by: Will Lahti <wtlahti@us.ibm.com>

tock-ibm commented on orderer/common/multichannel/registrar.go:
```diff
@@ -577,8 +577,56 @@ func (r *Registrar) joinAsFollower(ledgerRes *ledgerResources, clusterConsenter
+}
+
+func (r *Registrar) removeFollower(channelID string, follower *follower.Chain) error {
+	follower.Halt()
+
```
If the follower is doing on-boarding, we have to remove the join-block as well. If we don't, after a restart the orderer will rejoin... that's probably just a TODO comment at this stage, but thought I'd mention it so it won't get forgotten.

wlahti replied:
Added the TODO.

tock-ibm commented on integration/raft/channel_participation_test.go:
```diff
@@ -124,43 +125,89 @@ var _ = Describe("ChannelParticipation", func() {
+			By("listing all channels for orderer1")
 			channelparticipation.List(network, orderer1, []string{"participation-trophy", "another-participation-trophy"})
+
+			By("removing orderer1 from the consenter set")
+			channelConfig := nwo.GetConfig(network, peer, orderer2, "participation-trophy")
```
This is indeed the recommended flow. First remove from the consenters set, then remove from orderer. However - I have two comments:
1. The reverse flow should be working as well, you just have to wait for the remaining consenters to re-elect a leader in case the leader was the one removed.
2. When a raft chain detects eviction, it should turn into a follower. This is not yet implemented, the chain will just print something and exit. After https://jira.hyperledger.org/browse/FAB-17915 that will change.  So when that happens one should also check that the orderer behaves as expected when evicted from a channel.

I would insert a TODO and/or a Jira in the integration test story to catch these two test cases and cover then later.

wlahti replied:
Sounds good. I'lll open sub-tasks in the integration test task.

tock-ibm commented:
Other than the two minor comments about future work above, this looks good.
